design-proposals/passt-feature-gate-iface-api-deprecation.md

```diff
@@ -0,0 +1,176 @@
+and introduction of Passt network binding plugin [2](https://github.com/kubevirt/kubevirt/pull/10425) [3](https://github.com/kubevirt/kubevirt/pull/10621), 
+it's now possible to remove Passt network binding implementation from Kubevirt core.
+
+This proposal outlines the replacement plan for the existing Passt implementation and API.
+
+# Motivation
+Removing Passt network binding implementation from Kubevirt core reduces the maintenance burden.
+Having Passt binding code outside Kubevirt's core may ease contribution process as there is no need to be familiar 
+with [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt) core code.
+
+## Goals
+- Reduce maintenance of Passt network binding in Kubevirt core.
+  
+## Non Goals
+
+## User Stories
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt):
+- [Passt binding plugin hook sidecar](https://github.com/kubevirt/kubevirt/tree/main/cmd/sidecars/network-passt-binding)
+- [Passt binding plugin CNI](https://github.com/kubevirt/kubevirt/tree/main/cmd/cniplugins/passt-binding)
+
+# Design
+Replace Kubevirt core Passt network binding implementation and interface API with Passt binding plugin.
+Users should be aware of the coming changes and be prepared.
+
+The proposed plan for replacing Kubevirt core Passt binding implementation and API is as follows:
+
+1. Announce the coming changes for Passt binding described in this section at kubevirt-dev mailing list.
+   This will give users some time to prepare for changes in Kubevirt next release v1.2.0. 
+
+2. Kubevirt next release v1.2.0 should have the following changes:
+   1. Raise warning when Passt feature gate is enabled saying its deprecated.
+   2. Raise warning when Passt interface API is being used saying its deprecated and Passt binding plugin should be used instead.
+   3. Mark relevant filed as deprecated.
+
+   VM that use Passt interface API will work regularly.
+ 
+3. Kubevirt next release v1.3.0 should have the following changes:
+   1. [Passt interface API](http://kubevirt.io/api-reference/main/definitions.html#_v1_interfacepasst) 
+      usage will be blocked by the admission webhook.
+      Creating new VMs that use Passt interface API will be blocked.
+      Existing VMs that use Passt interface API won't manage to re-start.
+    
+      Users who like to use Passt binding will have to follow the network binding plugin API:
+      1. Install Passt binding CNI (make sure the binary available on all nodes).
+      2. Create and register Passt binding network-attachment-definition.
+      3. Register Passt binding sidecar image.
+         
+   2. Remove Passt binding implementation from Kubevirt core.
+   
+   3. Deprecate Passt feature gate.
+      Passt feature gate is set as no-op by being enabled by default, following Kubeivrt's [deprecation doc](https://github.com/kubevirt/kubevirt/blob/main/docs/deprecation.md).
```

Comment:
This will not work.
FG needs to be set hard-coded to disabled and not be editable anymore.

Reply:
Since passt implementation is going to be removed Passt FG will have no meaning, whether its enabled or not, isnt it?

Reply:
No, as this may have effect on old vm/s. But the definition is clear, when a feature is removed, it meams its fg is false and cannot be enabled anymore.

Reply:
How would it affect old VMs?
Running VMs will have no disruption, regardless of the FG state.

Reply:
I know of scenarios where the fg is passed to the virt-launcher. It may not effect passt feature, but it may effect other features.
In general, if a feature is discontinued, my interpretation is that its fg is false.
If the feature graduated to ga, it is true.

I can even inagine cases where the code is not changed at all, only the fg is being locked to one of the options.

Reply:
> I know of scenarios where the fg is passed to the virt-launcher. It may not effect passt feature, but it may effect other features.

Sorry I dont fully understand what you mean.
There are two FG that get passed to virt-launcher at syncVMI routine,non of them is Passt FG.
How deprecating Passt FG, in other words makes it always enabled,  is it going to affect other features?

> In general, if a feature is discontinued, my interpretation is that its fg is false.
If the feature graduated to ga, it is true.

The [deprecation doc](https://github.com/kubevirt/kubevirt/blob/main/docs/deprecation.md) lack in regard to discontinued features, there is on going discussion at https://github.com/kubevirt/kubevirt/issues/7745 though.

Passt FG refereed to only at admission webhook when it condition Passt FG to be enabled when using Passt interface API.
As soon as Passt interface API usage is always blocked, Passt FG has no meaning whether its enabled or not.

Is you insist of the semantics, where Passt is discontinued feature and shall be always set to false, we may need to extend https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-config/feature-gates.go#L96-L119
to take in account such features.
Althogtgh I dont think its worth it because Passt FG will have no affect on the code.